src/meta/processors/admin/ListClusterInfoProcessor.cpp

```diff
@@ -22,52 +22,98 @@ void ListClusterInfoProcessor::process(const cpp2::ListClusterInfoReq& req) {
     return;
   }
 
-  const auto& prefix = MetaKeyUtils::hostPrefix();
-  std::unique_ptr<kvstore::KVIterator> iter;
-  auto ret = kvstore_->prefix(kDefaultSpaceId, kDefaultPartId, prefix, &iter, true);
-  if (ret != nebula::cpp2::ErrorCode::SUCCEEDED) {
-    LOG(ERROR) << "prefix failed:" << apache::thrift::util::enumNameSafe(ret);
+  // services(include agent service) group by ip/hostname
+  std::unordered_map<std::string, std::vector<cpp2::ServiceInfo>> hostServices;
+
+  // non-meta services, may include inactive services
+  const auto& hostPrefix = MetaKeyUtils::hostPrefix();
+  auto iterRet = doPrefix(hostPrefix);
+  if (!nebula::ok(iterRet)) {
+    LOG(ERROR) << "get host prefix failed:"
+               << apache::thrift::util::enumNameSafe(nebula::error(iterRet));
     handleErrorCode(nebula::cpp2::ErrorCode::E_LIST_CLUSTER_FAILURE);
     onFinished();
     return;
   }
-
-  std::vector<nebula::cpp2::NodeInfo> storages;
-  while (iter->valid()) {
-    auto host = MetaKeyUtils::parseHostKey(iter->key());
+  auto iter = nebula::value(iterRet).get();
+  for (; iter->valid(); iter->next()) {
+    HostAddr addr = MetaKeyUtils::parseHostKey(iter->key());
     HostInfo info = HostInfo::decode(iter->val());
 
-    if (info.role_ != cpp2::HostRole::STORAGE) {
-      iter->next();
-      continue;
-    }
+    cpp2::ServiceInfo service;
+    service.set_role(info.role_);
+    service.set_addr(addr);
 
-    auto status = client_->listClusterInfo(host).get();
-    if (!status.ok()) {
-      LOG(ERROR) << "listcluster info from storage failed, host: " << host;
-      handleErrorCode(nebula::cpp2::ErrorCode::E_LIST_CLUSTER_FAILURE);
-      onFinished();
-      return;
+    // fill the dir info
+    if (info.role_ != meta::cpp2::HostRole::AGENT) {
```

Comment:
Maybe you should judge by `== GRAPH/META/STORAGE`, otherwise `LISTENER` is included, I am not sure what will happen...

Reply:
You are right, because when role== listener, I did not report it's dir info in the heartbeat.